libcontainer/container_linux.go

```diff
@@ -212,21 +212,46 @@ func (c *Container) Start(process *Process) error {
 // after start returns but opens the fifo after start returns.
 func (c *Container) Run(process *Process) error {
 	c.m.Lock()
-	defer c.m.Unlock()
 	if err := c.start(process); err != nil {
+		c.m.Unlock()
 		return err
 	}
 	if process.Init {
-		return c.exec()
+		err := c.exec()
+		if err != nil {
+			c.m.Unlock()
+			return err
+		}
+		c.m.Unlock()
+		return c.PostStart()
+	}
+	c.m.Unlock()
+	return nil
+}
+
+func (c *Container) PostStart() error {
+	s, err := c.OCIState()
```

Comment:
There is a lock in `c.OCIState()`, so you can use `c.currentOCIState()` instead.